### PR TITLE
Changelog 125

### PR DESCRIPTION
**Pull request by mchernyavsky**

### CODE REVIEW

File: _posts/2020-06-25-changelog-125.markdown [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+    TODO: GIF
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5609 -->
+* Update icons for Rust items
+
+    TODO: PNG (old / new)?
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/4640 -->
+* Enable intelligent highlighting inside macro call bodies
```
Comment (vlad20012) [inline _posts/2020-06-25-changelog-125.markdown:46]: Let's note that it works only with the new expansion engine and add instructions about how to enable it.

File: _posts/2020-06-25-changelog-125.markdown [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+* Support for `forbid` attribute (by [@Stzx])
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5313 -->
+* Allow the user to change mutability of parameters in `Extract Function` dialog (by [@Kobzol])
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5548 -->
+* Report type errors for references with different mutability even these references contain unknown types (by [@Kobzol])
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5570 -->
+* Make `Self` inside impl of enum to be resolved to both impl and enum items (by [@fan-tom])
```
Comment (vlad20012) [inline _posts/2020-06-25-changelog-125.markdown:74]: I'd say it's a fix

File: _posts/2020-06-25-changelog-125.markdown [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+  <!-- https://github.com/intellij-rust/intellij-rust/pull/5590 -->
+  * In structure view and usage tooltip
+  <!-- https://github.com/intellij-rust/intellij-rust/pull/5596 -->
+  * When formatting file with rustfmt on save
+
+## Internal Improvements
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5545 -->
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5506 -->
+* Now Rust plugin is [dynamically loadable](https://www.jetbrains.org/intellij/sdk/docs/basics/plugin_structure/dynamic_plugins.html) on 2020.2.
```
Comment (vlad20012) [inline _posts/2020-06-25-changelog-125.markdown:160]: Let's clarify that the plugin now can be installed without a restart. And I'd say it's a feature

File: _posts/2020-06-25-changelog-125.markdown [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5535 -->
+* Support debugging with MSVC toolchain in CLion 2020.2
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5586 -->
+* Add `Introduce local variable` intention (by [@Kobzol])
+
+    TODO: GIF
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5626 -->
+* Implement `Move File` refactoring
```
Comment (Undin) [inline _posts/2020-06-25-changelog-125.markdown:19]: Let add that this support is initial and add current restrictions

File: _posts/2020-06-25-changelog-125.markdown [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/4640 -->
+* Enable intelligent highlighting inside macro call bodies
+
+* Add detection of the new compiler errors:
+  <!-- https://github.com/intellij-rust/intellij-rust/pull/5375 -->
+  * You tried to use a type which doesn't implement some trait in a place which expected that trait [E0277](https://doc.rust-lang.org/error-index.html#E0277) (by [@Kobzol])
+  <!-- https://github.com/intellij-rust/intellij-rust/pull/5436 -->
+  * A constant item was initialized with something that is not a constant expression [E0015](https://doc.rust-lang.org/error-index.html#E0015) (by [@t-kameyama])
+  * A non-constant value was used in a constant expression [E0435](https://doc.rust-lang.org/error-index.html#E0435) (by [@t-kameyama])
```
Comment (Undin) [inline _posts/2020-06-25-changelog-125.markdown:53]: I'd combine this point with previous one and clarify that it's only for array size expressions
Something like `Check when a non-constant item is used in array size expressions [E0015] & [E0435]`

File: _posts/2020-06-25-changelog-125.markdown [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+  * In structure view and usage tooltip
+  <!-- https://github.com/intellij-rust/intellij-rust/pull/5596 -->
+  * When formatting file with rustfmt on save
+
+## Internal Improvements
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5545 -->
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5506 -->
+* Now Rust plugin is [dynamically loadable](https://www.jetbrains.org/intellij/sdk/docs/basics/plugin_structure/dynamic_plugins.html) on 2020.2.
+It allows you to install and update the plugin without restarting the IDE
```
Comment (Undin) [inline _posts/2020-06-25-changelog-125.markdown:161]: Let's avoid `update` here because update should unload plugin first and it may not work

File: _posts/2020-06-25-changelog-125.markdown (hunk 3 above)
Comment (Undin) [reply _posts/2020-06-25-changelog-125.markdown:160]: > Let's clarify that the plugin now can be installed without a restart.

It's mentioned below

> And I'd say it's a feature

Agree

File: _posts/2020-06-25-changelog-125.markdown [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+* Fix `Join Lines` action for doc comment lines (by [@Stzx])
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5515 -->
+* Disable `Promote Module to Directory` and `Downgrade Module to File` refactorings for invalid files in the editor (by [@Kobzol])
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/4868 -->
+* Correctly handle symlinked Cargo projects (by [@Jezza])
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5603 -->
+* Take into account namespaces when name shadowing occurs
```
Comment (Undin) [inline _posts/2020-06-25-changelog-125.markdown:124]: Let's mentioned what it means for user, i.e. name resolution now works when user use `anyhow::anyhow` uses

Comment (Undin) [inline _posts/2020-06-25-changelog-125.markdown:124]: Also, I think it should be at the top of fixes

File: _posts/2020-06-25-changelog-125.markdown [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+* Allow the user to change mutability of parameters in `Extract Function` dialog (by [@Kobzol])
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5548 -->
+* Report type errors for references with different mutability even these references contain unknown types (by [@Kobzol])
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5570 -->
+* Make `Self` inside impl of enum to be resolved to both impl and enum items (by [@fan-tom])
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5580 -->
+* Enable `Convert match statement to if let` intention for irrefutable patterns (by [@Kobzol])
```
Comment (Undin) [inline _posts/2020-06-25-changelog-125.markdown:77]: Let's add link https://doc.rust-lang.org/book/ch18-02-refutability.html for "irrefutable patterns"

File: _posts/2020-06-25-changelog-125.markdown [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+* Support for `non_exhaustive` attribute (by [@Mcat12])
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5571 -->
+* Support for `forbid` attribute (by [@Stzx])
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5313 -->
+* Allow the user to change mutability of parameters in `Extract Function` dialog (by [@Kobzol])
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5548 -->
+* Report type errors for references with different mutability even these references contain unknown types (by [@Kobzol])
```
Comment (Undin) [inline _posts/2020-06-25-changelog-125.markdown:71]:
```suggestion
* Report type errors for references with different mutability even if a type is not fully known (by [@Kobzol])
```

File: _posts/2020-06-25-changelog-125.markdown [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+  * An unrecognized representation attribute was used [E0552](https://doc.rust-lang.org/error-index.html#E0552) (by [@Stzx])
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5571 -->
+* Now IDE's inspections take into account lint level attributes (by [@Stzx])
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5466 -->
+* Support for `non_exhaustive` attribute (by [@Mcat12])
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5571 -->
+* Support for `forbid` attribute (by [@Stzx])
```
Comment (Undin) [inline _posts/2020-06-25-changelog-125.markdown:65]: I suppose it's quite hard for users to understand what it means. Let's reword it to clarify that exactly we support

File: _posts/2020-06-25-changelog-125.markdown [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+  * You tried to use a type which doesn't implement some trait in a place which expected that trait [E0277](https://doc.rust-lang.org/error-index.html#E0277) (by [@Kobzol])
+  <!-- https://github.com/intellij-rust/intellij-rust/pull/5436 -->
+  * A constant item was initialized with something that is not a constant expression [E0015](https://doc.rust-lang.org/error-index.html#E0015) (by [@t-kameyama])
+  * A non-constant value was used in a constant expression [E0435](https://doc.rust-lang.org/error-index.html#E0435) (by [@t-kameyama])
+  <!-- https://github.com/intellij-rust/intellij-rust/pull/5582 -->
+  * A representation attribute was placed on an unsupported item [E0517](https://doc.rust-lang.org/error-index.html#E0517) (by [@Stzx])
+  * An unrecognized representation attribute was used [E0552](https://doc.rust-lang.org/error-index.html#E0552) (by [@Stzx])
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5571 -->
+* Now IDE's inspections take into account lint level attributes (by [@Stzx])
```
Comment (Undin) [inline _posts/2020-06-25-changelog-125.markdown:59]:
```suggestion
* Now plugin's inspections take into account lint level attributes (by [@Stzx])
```

Comment (Undin) [reply _posts/2020-06-25-changelog-125.markdown:59]: Would be great to specify what `take into account` means for users

File: _posts/2020-06-25-changelog-125.markdown [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+    TODO: GIF
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5557 -->
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5561 -->
+* Add quick fixes to remove unused variable / parameter (by [@Kobzol])
+
+    TODO: GIF
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5609 -->
+* Update icons for Rust items
```
Comment (Undin) [inline _posts/2020-06-25-changelog-125.markdown:41]:
```suggestion
* Update icons for Rust items. Now they are more similar to C/C++ icons
```

File: _posts/2020-06-25-changelog-125.markdown [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+* Enable intelligent highlighting inside macro call bodies
+
+* Add detection of the new compiler errors:
+  <!-- https://github.com/intellij-rust/intellij-rust/pull/5375 -->
+  * You tried to use a type which doesn't implement some trait in a place which expected that trait [E0277](https://doc.rust-lang.org/error-index.html#E0277) (by [@Kobzol])
+  <!-- https://github.com/intellij-rust/intellij-rust/pull/5436 -->
+  * A constant item was initialized with something that is not a constant expression [E0015](https://doc.rust-lang.org/error-index.html#E0015) (by [@t-kameyama])
+  * A non-constant value was used in a constant expression [E0435](https://doc.rust-lang.org/error-index.html#E0435) (by [@t-kameyama])
+  <!-- https://github.com/intellij-rust/intellij-rust/pull/5582 -->
+  * A representation attribute was placed on an unsupported item [E0517](https://doc.rust-lang.org/error-index.html#E0517) (by [@Stzx])
```
Comment (Undin) [inline _posts/2020-06-25-changelog-125.markdown:55]:
```suggestion
  * A `[repr](https://doc.rust-lang.org/reference/type-layout.html#representations)` attribute is placed on an unsupported item [E0517](https://doc.rust-lang.org/error-index.html#E0517) (by [@Stzx])
```

File: _posts/2020-06-25-changelog-125.markdown [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+* Update icons for Rust items
+
+    TODO: PNG (old / new)?
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/4640 -->
+* Enable intelligent highlighting inside macro call bodies
+
+* Add detection of the new compiler errors:
+  <!-- https://github.com/intellij-rust/intellij-rust/pull/5375 -->
+  * You tried to use a type which doesn't implement some trait in a place which expected that trait [E0277](https://doc.rust-lang.org/error-index.html#E0277) (by [@Kobzol])
```
Comment (Undin) [inline _posts/2020-06-25-changelog-125.markdown:50]: Let's use a passive voice here

File: _posts/2020-06-25-changelog-125.markdown [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+  * A non-constant value was used in a constant expression [E0435](https://doc.rust-lang.org/error-index.html#E0435) (by [@t-kameyama])
+  <!-- https://github.com/intellij-rust/intellij-rust/pull/5582 -->
+  * A representation attribute was placed on an unsupported item [E0517](https://doc.rust-lang.org/error-index.html#E0517) (by [@Stzx])
+  * An unrecognized representation attribute was used [E0552](https://doc.rust-lang.org/error-index.html#E0552) (by [@Stzx])
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5571 -->
+* Now IDE's inspections take into account lint level attributes (by [@Stzx])
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5466 -->
+* Support for `non_exhaustive` attribute (by [@Mcat12])
```
Comment (Undin) [inline _posts/2020-06-25-changelog-125.markdown:62]: But it's not true - `non_exhaustive ` attribute had been supported in `Match check` inspection before this fix.
The corresponding PR fixes false positive "Match must be exhaustive" error

File: _posts/2020-06-25-changelog-125.markdown [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+* Show unique editor tab names for `main.rs`, `lib.rs` and `Cargo.toml` (by [@Kobzol])
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5533 -->
+* Improve text for `Convert to type` quick fix (by [@Kobzol])
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5354 -->
+* Update `Wrong type arguments number` inspection (by [@Kobzol])
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5462 -->
+* Improve `Remove redundant type arguments` quick fix (by [@Stzx])
```
Comment (Undin) [inline _posts/2020-06-25-changelog-125.markdown:89]: I don't understand what exactly we did here from end-user point of view

File: _posts/2020-06-25-changelog-125.markdown [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+
+* Add detection of the new compiler errors:
+  <!-- https://github.com/intellij-rust/intellij-rust/pull/5375 -->
+  * You tried to use a type which doesn't implement some trait in a place which expected that trait [E0277](https://doc.rust-lang.org/error-index.html#E0277) (by [@Kobzol])
+  <!-- https://github.com/intellij-rust/intellij-rust/pull/5436 -->
+  * A constant item was initialized with something that is not a constant expression [E0015](https://doc.rust-lang.org/error-index.html#E0015) (by [@t-kameyama])
+  * A non-constant value was used in a constant expression [E0435](https://doc.rust-lang.org/error-index.html#E0435) (by [@t-kameyama])
+  <!-- https://github.com/intellij-rust/intellij-rust/pull/5582 -->
+  * A representation attribute was placed on an unsupported item [E0517](https://doc.rust-lang.org/error-index.html#E0517) (by [@Stzx])
+  * An unrecognized representation attribute was used [E0552](https://doc.rust-lang.org/error-index.html#E0552) (by [@Stzx])
```
Comment (Undin) [inline _posts/2020-06-25-changelog-125.markdown:56]: Maybe merge all error detections related to repr attribute into one item?

Comment (Undin) [reply _posts/2020-06-25-changelog-125.markdown:56]: Hm, is it worth to duplicate error description since we produce the corresponding link to error index?

File: _posts/2020-06-25-changelog-125.markdown [hunk 18]
```diff
@@ -0,0 +1,176 @@
...
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5609 -->
+* Update icons for Rust items
+
+    TODO: PNG (old / new)?
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/4640 -->
+* Enable intelligent highlighting inside macro call bodies
+
+* Add detection of the new compiler errors:
```
Comment (Undin) [inline _posts/2020-06-25-changelog-125.markdown:48]: Let's use present tense instead of past one in items below

File: _posts/2020-06-25-changelog-125.markdown [hunk 19]
```diff
@@ -0,0 +1,176 @@
...
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5462 -->
+* Improve `Remove redundant type arguments` quick fix (by [@Stzx])
+
+## Fixes
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5599 -->
+* Fix false positive about not allowed `impl Trait` in associated type binding (by [@Kobzol])
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5381 -->
+* Fix usage replacement in the `Extract Enum Variant` intention (by [@Kobzol])
```
Comment (Undin) [inline _posts/2020-06-25-changelog-125.markdown:97]:
```suggestion
* Fix usage replacement in the `Extract Enum Variant` refactoring (by [@Kobzol])
```

File: _posts/2020-06-25-changelog-125.markdown [hunk 20]
```diff
@@ -0,0 +1,176 @@
...
+layout: post
+title: "IntelliJ Rust Changelog #125"
+date: 2020-06-25 19:26:11 +0300
+---
+
+
+## New Features
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5535 -->
+* Support debugging with MSVC toolchain in CLion 2020.2
```
Comment (Undin) [inline _posts/2020-06-25-changelog-125.markdown:11]: Let's also mention https://github.com/intellij-rust/intellij-rust/issues/5632 as the current status of debugger for msvc toolchain

File: _posts/2020-06-25-changelog-125.markdown [hunk 21]
```diff
@@ -0,0 +1,173 @@
+---
+layout: post
+title: "IntelliJ Rust Changelog #125"
+date: 2020-06-25 19:26:11 +0300
```
Comment (Undin) [inline _posts/2020-06-25-changelog-125.markdown:4]: Don't forget to update date

File: _posts/2020-06-29-changelog-125.markdown [hunk 22]
```diff
@@ -0,0 +1,173 @@
...
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5603 -->
+* Take into account namespaces when name shadowing occurs.
+For example, now `use anyhow::anyhow` doesn't break name resolution
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5599 -->
+* Fix false positive about not allowed `impl Trait` in associated type binding (by [@Kobzol])
+
+<!-- https://github.com/intellij-rust/intellij-rust/pull/5466 -->
+* Fix false positive for "Match must be exhaustive" error (by [@Mcat12])
```
Comment (Undin) [inline _posts/2020-06-29-changelog-125.markdown:96]:
```suggestion
* Fix false positive for `Match must be exhaustive` error (by [@Mcat12])
```

**Comment by Undin** [comment]
Great!
So huge release :)
